### PR TITLE
Fix undefined behavior in copy_blocks when source and destination blocks overlap

### PR DESCRIPTION
**Pull request by yafshar**
### Summary
`copy_blocks` uses `index_copy_` with overlapping source and destination indices, which is explicitly undefined behavior in PyTorch.
On HPU/GPU devices this can silently corrupt the KV cache when the same physical block appears more than once in the `block_mapping` list.
The patch switches to a gather-then-scatter sequence (`index_select` + `index_put_`) so that every source block is read before any destination slot is overwritten.

#### Changes
`copy_blocks`: introduce temporary tensors to avoid read-after-write hazards

#### Issue reproduction
The below unit test fails on main branch and passes after the fix.

```py
import torch
import habana_frameworks.torch as htorch


def copy_blocks_new(key_caches, value_caches, block_mapping):
    if block_mapping.numel() == 0:
        return

    block_mapping = block_mapping.transpose(0, 1)
    src = block_mapping[0]
    dst = block_mapping[1]

    for key_cache, value_cache in zip(key_caches, value_caches):
        # read once, write once - no overlap
        k_values = key_cache.index_select(0, src)   # gather
        v_values = value_cache.index_select(0, src)
        key_cache.index_put_([dst], k_values)       # scatter
        value_cache.index_put_([dst], v_values)


def copy_blocks_old(key_caches, value_caches, block_mapping):
    if block_mapping.numel() == 0:
        return

    block_mapping = block_mapping.transpose(0, 1)
    src = block_mapping[0]
    dst = block_mapping[1]

    for key_cache, value_cache in zip(key_caches, value_caches):
        key_cache.index_copy_(0, dst, key_cache.index_select(0, src))
        value_cache.index_copy_(0, dst, value_cache.index_select(0, src))

    if key_caches[0].device.type == 'hpu':
        htorch.core.mark_step()


def test_copy(fn, name):
    num_blocks, block_size, num_heads, head_size = 4, 128, 8, 64
    k = torch.arange(num_blocks * block_size * num_heads * head_size,
                     dtype=torch.bfloat16, device='hpu').view(num_blocks, num_heads, block_size, head_size)
    v = k + 1000

    mapping = torch.tensor([[2, 0], [1, 0]], dtype=torch.long, device='hpu')

    orig_k = k.clone().cpu()
    orig_v = v.clone().cpu()

    fn([k], [v], mapping)           # run the function under test
    torch.hpu.synchronize()

    dst = 0
    src = 1

    try:
        torch.testing.assert_close(k[dst].cpu(), orig_k[src])
        torch.testing.assert_close(v[dst].cpu(), orig_v[src])
        print(f"�~\~T {name}: overlap chain copied correctly")
    except AssertionError as e:
        print(f"�~\~X {name}: FAILED �~@~S {e}")

#  run both
test_copy(copy_blocks_old, "OLD (UB)")
test_copy(copy_blocks_new, "NEW (safe)")
```

Running the test.py
```sh
>>> python test.py
✘ OLD (UB): FAILED – Tensor-likes are not close!

Mismatched elements: 21376 / 65536 (32.6%)
Greatest absolute difference: 66048.0 at index (0, 8, 1) (up to 1e-05 allowed)
Greatest relative difference: 1.0 at index (0, 4, 0) (up to 0.016 allowed)

✔ NEW (safe): overlap chain copied correctly
```

### CODE REVIEW

**Comment by xuechendi** [comment]
@kzawora-intel  @madamczyk-intel  @mswiniarsk , may you check this PR?

I updated original UT with hpu graph so we can test on what actually happened during model.fwd

with out HPU_graph, the statement in this PR is true
```
PT_HPU_LAZY_MODE=1  python examples/test_index_copy.py
❌ OLD (UB): FAILED
   Tensor-likes are not close!

Mismatched elements: 21376 / 65536 (32.6%)
Greatest absolute difference: 66048.0 at index (0, 9, 0) (up to 1e-05 allowed)
Greatest relative difference: 1.0 at index (0, 0, 0) (up to 0.016 allowed)
OLD copy - run1 took 3.027279 seconds
❌ OLD (UB): FAILED
   Tensor-likes are not close!

Mismatched elements: 26880 / 65536 (41.0%)
Greatest absolute difference: 66048.0 at index (0, 8, 1) (up to 1e-05 allowed)
Greatest relative difference: 1.0 at index (0, 0, 0) (up to 0.016 allowed)
OLD copy - run2 took 0.003058 seconds
========================================
✅ NEW (safe): Overlap chain copied correctly
NEW copy - run1 took 0.008351 seconds
✅ NEW (safe): Overlap chain copied correctly
NEW copy - run2 took 0.002630 seconds
```


Somehow, with HPU_graph, both old copy and new copy failed on data integrity check
```
PT_HPU_LAZY_MODE=1  python examples/test_index_copy.py --use_hpu_graph

mpatibleWithHPUGraphs)
❌ OLD (UB): FAILED
   Tensor-likes are not close!

Mismatched elements: 25792 / 65536 (39.4%)
Greatest absolute difference: 66048.0 at index (0, 9, 0) (up to 1e-05 allowed)
Greatest relative difference: 1.0 at index (0, 0, 0) (up to 0.016 allowed)
OLD copy - run1 took 0.116837 seconds
❌ OLD (UB): FAILED
   Tensor-likes are not close!

Mismatched elements: 65536 / 65536 (100.0%)
Greatest absolute difference: 65536.0 at index (0, 0, 0) (up to 1e-05 allowed)
Greatest relative difference: 1.0 at index (0, 0, 0) (up to 0.016 allowed)
OLD copy - run2 took 0.005335 seconds
========================================
✅ NEW (safe): Overlap chain copied correctly
NEW copy - run1 took 0.008230 seconds
❌ NEW (safe): FAILED
   Tensor-likes are not close!

Mismatched elements: 65536 / 65536 (100.0%)
Greatest absolute difference: 65536.0 at index (0, 0, 0) (up to 1e-05 allowed)
Greatest relative difference: 1.0 at index (0, 0, 0) (up to 0.016 allowed)
NEW copy - run2 took 0.002626 seconds
```

```
import argparse
import os
import time
import torch
import habana_frameworks.torch as htorch

class CacheCopierOld(torch.nn.Module):
    """
    The original copy logic, now encapsulated in a torch.nn.Module.
    """
    def __init__(self):
        super().__init__()

    def forward(self, key_caches, value_caches, block_mapping):
        if block_mapping.numel() == 0:
            return

        block_mapping = block_mapping.transpose(0, 1)
        src = block_mapping[0]
        dst = block_mapping[1]

        for key_cache, value_cache in zip(key_caches, value_caches):
            # This operation has undefined behavior (UB) if src and dst overlap.
            key_cache.index_copy_(0, dst, key_cache.index_select(0, src))
            value_cache.index_copy_(0, dst, value_cache.index_select(0, src))


class CacheCopierNew(torch.nn.Module):
    """
    The updated, safe copy logic, now encapsulated in a torch.nn.Module.
    """
    def __init__(self):
        super().__init__()

    def forward(self, key_caches, value_caches, block_mapping):
        if block_mapping.numel() == 0:
            return

        block_mapping = block_mapping.transpose(0, 1)
        src = block_mapping[0]
        dst = block_mapping[1]

        for key_cache, value_cache in zip(key_caches, value_caches):
            # Safe implementation: gather all source values first, then scatter to destination.
            # This avoids read/write overlaps.
            k_values = key_cache.index_select(0, src)  # gather
            v_values = value_cache.index_select(0, src)
            key_cache.index_put_([dst], k_values)      # scatter
            value_cache.index_put_([dst], v_values)


def test_copy(model_instance, name):
    """
    Test function now accepts a model instance (torch.nn.Module) and calls it.
    """
    num_blocks, block_size, num_heads, head_size = 4, 128, 8, 64
    k = torch.arange(num_blocks * block_size * num_heads * head_size,
                     dtype=torch.bfloat16, device='hpu').view(num_blocks, num_heads, block_size, head_size)
    v = k + 1000

    # Test case with overlapping source and destination indices in the chain: 1 -> 0, 2 -> 0
    mapping = torch.tensor([[2, 0], [1, 0]], dtype=torch.long, device='hpu')

    orig_k = k.clone().cpu()
    orig_v = v.clone().cpu()

    # Run the model's forward pass
    model_instance(key_caches=[k], value_caches=[v], block_mapping=mapping)
    torch.hpu.synchronize()

    dst = 0
    # In the mapping [[2, 0], [1, 0]], the copy for index 1 happens last.
    # Therefore, the final state of dst=0 should match the original state of src=1.
    src = 1

    try:
        torch.testing.assert_close(k[dst].cpu(), orig_k[src])
        torch.testing.assert_close(v[dst].cpu(), orig_v[src])
        print(f"✅ {name}: Overlap chain copied correctly")
    except AssertionError as e:
        print(f"❌ {name}: FAILED\n   {e}")


# --- Main execution ---
if __name__ == "__main__":

    print("Starting cache copy tests...\n")
    argparser = argparse.ArgumentParser()
    argparser.add_argument("--use_hpu_graph", action="store_true", help="Use HPU graph")
    args = argparser.parse_args()

    # Instantiate and run the old model
    copier_old = CacheCopierOld()
    if args.use_hpu_graph:
        copier_old = htorch.hpu.wrap_in_hpu_graph(copier_old)
    start = time.perf_counter()
    test_copy(copier_old, "OLD (UB)")
    elapsed = time.perf_counter() - start
    print(f"OLD copy - run1 took {elapsed:.6f} seconds")

    start = time.perf_counter()
    test_copy(copier_old, "OLD (UB)")
    elapsed = time.perf_counter() - start
    print(f"OLD copy - run2 took {elapsed:.6f} seconds")

    print("========================================")

    # Instantiate and run the new model
    copier_new = CacheCopierNew()
    if args.use_hpu_graph:
        copier_new = htorch.hpu.wrap_in_hpu_graph(copier_new)
    start = time.perf_counter()
    test_copy(copier_new, "NEW (safe)")
    elapsed = time.perf_counter() - start
    print(f"NEW copy - run1 took {elapsed:.6f} seconds")

    start = time.perf_counter()
    test_copy(copier_new, "NEW (safe)")
    elapsed = time.perf_counter() - start
    print(f"NEW copy - run2 took {elapsed:.6f} seconds")
```

**Comment by xuechendi** [comment]
@yafshar , I am also not sure what use case will lead to `mapping = torch.tensor([[2, 0], [1, 0]], dtype=torch.long, device='hpu')` scenario?

In your proposed scenario, it seems to me that there is two tokens update to the same KV slot?  Why this happens?

**Comment by xuechendi** [comment]
I see, thanks for the explanation. 
Can you provide a complete configuration, the use case two seqs linked to same block is due to prefix_cache right?

**Comment by yafshar** [comment]
@xuechendi would you please check this PR gain.
We encountered an FP8 prefix-caching issue in heterogeneous scenarios. The latest commit, along with another commit to enable GDR in the PR, resolved the accuracy problem for cases where the KV-cache in FP8 is transferred from NVIDIA to Gaudi. Using index_put_ in both eager mode and torch.compile caused accuracy issues, which we addressed by switching to index_copy_.

**Comment by adobrzyn** [comment]
@yafshar - is this PR still needed? Are You waiting for review?

**Comment by yafshar** [comment]
@xuechendi @michalkuligowski @adobrzyn Just following up on this PR! It addresses a clear bug related to prefix caching. Could you please review when you have a chance? Let me know if you need any additional details or changes. Thanks!
Overlapping source and destination indices, is explicitly undefined behavior in PyTorch and a clear bug!

**Comment by adobrzyn** [comment]
@yafshar - please resolve conficts to have branch up to date.

It looks reasonable, but have you tested this change somewhere? On some benchmarks and acc tests?

**Comment by yafshar** [comment]
> @yafshar - please resolve conficts to have branch up to date.
> 
> It looks reasonable, but have you tested this change somewhere? On some benchmarks and acc tests?

Yes, the code has been tested and is currently used in the heterogeneous HPU + GPU workflow (llm-d & Dynamo).
I didn’t write a separate unit test specifically for this function, but you can reproduce the issue using my earlier reproducer. This is a real bug and needs to be fixed. I’m not sure why @xuechendi hasn’t merged the fix yet.

**Review by michalkuligowski — APPROVED** [review]
